### PR TITLE
first draft of splitting NWSS signals

### PR DESCRIPTION
**Pull request by dsweber2**
### Description
This splits the dataset based on the provider and normalization (not every pair is actually present), and adds the metric signals. The resulting signals are called:
- pcr_conc_smoothed_CDC_VERILY_flow-population
- detect_prop_15d_CDC_VERILY_flow-population
- percentile_CDC_VERILY_flow-population
- ptc_15d_CDC_VERILY_flow-population
- pcr_conc_smoothed_CDC_VERILY_microbial
- detect_prop_15d_CDC_VERILY_microbial
- percentile_CDC_VERILY_microbial
- ptc_15d_CDC_VERILY_microbial
- pcr_conc_smoothed_NWSS_flow-population
- detect_prop_15d_NWSS_flow-population
- percentile_NWSS_flow-population
- ptc_15d_NWSS_flow-population
- pcr_conc_smoothed_NWSS_microbial
- detect_prop_15d_NWSS_microbial
- percentile_NWSS_microbial
- ptc_15d_NWSS_microbial
- pcr_conc_smoothed_WWS_microbial
- detect_prop_15d_WWS_microbial
- percentile_WWS_microbial
- ptc_15d_WWS_microbial

Some of these can have negative values; for e.g. `ptc_15d`, the values are small enough that I expect these may actually be exponents. Still looking into why the concentration data has negative values, which are too large to make sense as exponents.

### Fixes 
- Fix `generate_weights` in the case where some weights are negative, and added a test for it

### CODE REVIEW

**Comment by dsweber2** [comment]
Minh confirmed that this runs on her machine and that the output looks reasonable. A couple of things to do before we merge this:
- [ ] figure out the right cut-off for sig digits so that the noise is removed
- [ ] seeing what values the formerly negative concentrations became
- [ ] write docs similar to [this PR](https://github.com/cmu-delphi/delphi-epidata/pull/1066)

**Review by dshemetov — APPROVED** [review]

**Comment by dsweber2** [comment]
Currently not passing because of the same update that made nssp tests fail. Once that's merged and this is rebased it will pass.

File: nwss_wastewater/delphi_nwss/run.py [hunk 1]
```diff
@@ -126,38 +78,52 @@ def run_module(params):
...
+        df_prov_norm = df_pull[(df_pull.provider == provider) & (df_pull.normalization == normalization)]
+        df_prov_norm = df_prov_norm.drop(["provider", "normalization"], axis=1)
+        for sensor in [*SIGNALS, *METRIC_SIGNALS]:
+            full_sensor_name = sensor + "_" + provider + "_" + normalization
+            df_prov_norm = df_prov_norm.rename(columns={sensor: full_sensor_name})
+            for geo in GEOS:
+                logger.info("Generating signal and exporting to CSV", metric=full_sensor_name)
+                if geo == "nation":
+                    df_prov_norm["nation"] = "us"
+                agg_df = geomapper.aggregate_by_weighted_sum(
```
Comment (nmdefries) [inline nwss_wastewater/delphi_nwss/run.py:107]: question: looks like we're aggregating for all geos (state and nation). Is the base geo type not reportable?

File: nwss_wastewater/delphi_nwss/run.py [hunk 2]
```diff
@@ -126,38 +78,52 @@ def run_module(params):
...
+                agg_df = add_needed_columns(agg_df)
+                # actual export
+                dates = create_export_csv(agg_df, geo_res=geo, export_dir=export_dir, sensor=full_sensor_name)
+                if "archive" in params:
+                    _, common_diffs, new_files = arch_diff.diff_exports()
+                    to_archive = [f for f, diff in common_diffs.items() if diff is not None]
+                    to_archive += new_files
+                    _, fails = arch_diff.archive_exports(to_archive)
+                    succ_common_diffs = {f: diff for f, diff in common_diffs.items() if f not in fails}
+                    arch_diff.filter_exports(succ_common_diffs)
```
Comment (nmdefries) [inline nwss_wastewater/delphi_nwss/run.py:125]: issue: same thing about [the runner script](https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/_delphi_utils_python/delphi_utils/runner.py)

File: nwss_wastewater/delphi_nwss/run.py [hunk 3]
```diff
@@ -126,38 +78,52 @@ def run_module(params):
...
+    df_pull = pull_nwss_data(socrata_token, logger)
+    geomapper = GeoMapper()
+    # iterate over the providers and the normalizations that they specifically provide
+    for provider, normalization in zip(PROVIDER_NORMS["provider"], PROVIDER_NORMS["normalization"]):
+        # copy by only taking the relevant subsection
+        df_prov_norm = df_pull[(df_pull.provider == provider) & (df_pull.normalization == normalization)]
+        df_prov_norm = df_prov_norm.drop(["provider", "normalization"], axis=1)
+        for sensor in [*SIGNALS, *METRIC_SIGNALS]:
+            full_sensor_name = sensor + "_" + provider + "_" + normalization
+            df_prov_norm = df_prov_norm.rename(columns={sensor: full_sensor_name})
```
Comment (nmdefries) [inline nwss_wastewater/delphi_nwss/run.py:102]: nit: I don't think this rename is necessary. We provide the field name during aggregation and then immediately change the name following that.

File: nwss_wastewater/delphi_nwss/run.py [hunk 4]
```diff
@@ -126,38 +78,52 @@ def run_module(params):
...
+        arch_diff = S3ArchiveDiffer(
             params["archive"]["cache_dir"],
             export_dir,
             params["archive"]["bucket_name"],
-            "nchs_mortality",
+            "nwss_wastewater",
             params["archive"]["aws_credentials"],
         )
-        daily_arch_diff.update_cache()
+        arch_diff.update_cache()
```
Comment (nmdefries) [inline nwss_wastewater/delphi_nwss/run.py:88]: issue: you don't need to run the archive differ, [the runner script](https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/_delphi_utils_python/delphi_utils/runner.py) takes care of that.

File: nwss_wastewater/delphi_nwss/constants.py [hunk 5]
```diff
@@ -12,18 +12,33 @@
 
 SIGNALS = ["pcr_conc_smoothed"]
 METRIC_SIGNALS = ["detect_prop_15d", "percentile", "ptc_15d"]
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply nwss_wastewater/delphi_nwss/constants.py:14]: suggestion: I find these signal names hard to parse. I'd prefer longer, more descriptive names. Our final signal names will include source and normalization method, though, so maybe they'd get too long?

Worth more thought. Have you checked these names with Roni yet?

**Review by nmdefries — CHANGES_REQUESTED** [review]
Some comments about style and comments. The actual functionality here looks fine.

This hasn't been released at all yet, right? Have we done statistical review on this?

**Comment by dsweber2** [comment]
> This hasn't been released at all yet, right? You previously did statistical review on this; is there anything else we want to do for these new signals?

It has not. There's the [corresponding docs](https://github.com/cmu-delphi/delphi-epidata/pull/1405), which I think Will read through at one point.

File: nwss_wastewater/delphi_nwss/constants.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/constants.py:14]: > What is the distinction between these two signal sets?

They're from two separate socrata APIs.

I haven't run the names by Roni, that's a good idea. The names are based on mirroring the original dataset's names.

File: nwss_wastewater/delphi_nwss/pull.py [hunk 6]
```diff
@@ -95,40 +111,43 @@ def pull_nwss_data(socrata_token: str):
...
         Dataframe as described above.
     """
-    # concentration key types
-    type_dict, type_dict_metric = construct_typedicts()
-
     # Pull data from Socrata API
-    client = Socrata("data.cdc.gov", socrata_token)
+    client = Socrata("data.cdc.gov", token)
     results_concentration = client.get("g653-rqe2", limit=10**10)
     results_metric = client.get("2ew6-ywp6", limit=10**10)
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:117]: do you mean `"data.cdc.gov"` and `"g653-rq32"`?

File: nwss_wastewater/delphi_nwss/constants.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nmdefries) [reply nwss_wastewater/delphi_nwss/constants.py:14]: It might have been left out of the "how to make an indicator" doc, but officially we're supposed to check signal names with Roni.

File: nwss_wastewater/delphi_nwss/pull.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nmdefries) [reply nwss_wastewater/delphi_nwss/pull.py:117]: Yep

File: nwss_wastewater/delphi_nwss/pull.py [hunk 7]
```diff
@@ -95,40 +111,43 @@ def pull_nwss_data(socrata_token: str):
...
-    df_concentration[SIGNALS[0]] = sig_digit_round(
-        df_concentration[SIGNALS[0]], SIG_DIGITS
-    )
+    for signal in [*SIGNALS, *METRIC_SIGNALS]:
+        df[signal] = sig_digit_round(df[signal], SIG_DIGITS)
 
-    df_concentration = add_population(df_concentration, df_metric)
     # if there are population NA's, assume the previous value is accurate (most
     # likely introduced by dates only present in one and not the other; even
     # otherwise, best to assume some value rather than break the data)
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:140]: It is; only one of them actually has population data listed ([the metric data](https://data.cdc.gov/Public-Health-Surveillance/NWSS-Public-SARS-CoV-2-Wastewater-Metric-Data/2ew6-ywp6/data_preview) vs [the concentration](https://data.cdc.gov/Public-Health-Surveillance/NWSS-Public-SARS-CoV-2-Concentration-in-Wastewater/g653-rqe2/data_preview)). There's actually a ton of signal metadata present in the metric data not in the concentration data.

File: nwss_wastewater/delphi_nwss/run.py (hunk 1 above)
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/run.py:107]: The base geo is `(state, wwtp_id)` (wastewater treatment plant id), which doesn't really match any other source at all. I was planning on waiting to add that, possibly outside of covidcast-indicators. Do you think just adding it to covidcast-indicators is a good idea?

File: nwss_wastewater/delphi_nwss/run.py (hunk 2 above)
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/run.py:125]: sorry I don't follow?

Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/run.py:125]: oh this was commented in the other order, I'll just delete this then. There are several endpoints which include this btw

File: nwss_wastewater/delphi_nwss/pull.py [hunk 8]
```diff
@@ -95,40 +111,43 @@ def pull_nwss_data(socrata_token: str):
...
-    df_concentration.population_served = df_concentration.population_served.ffill()
-
-    keep_columns = ["timestamp", "state", "population_served"]
-    return df_concentration[SIGNALS + keep_columns]
+    df.population_served = df.population_served.ffill()
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:141]: good catch; I'm adding a `sort_index` to the reformat function to address this (p sure the index is the right order).

File: nwss_wastewater/delphi_nwss/pull.py [hunk 9]
```diff
@@ -95,40 +111,43 @@ def pull_nwss_data(socrata_token: str):
...
-        df_metric = df_metric.astype(type_dict_metric)
-    except KeyError as exc:
-        raise ValueError(warn_string(df_metric, type_dict_metric)) from exc
+    # Drop sites without a normalization scheme.
+    df = df_concentration[~df_concentration["normalization"].isna()]
 
-    # pull 2 letter state labels out of the key_plot_id labels
-    df_concentration["state"] = df_concentration.key_plot_id.str.extract(r"_(\w\w)_")
+    # Pull 2 letter state labels out of the key_plot_id labels.
+    add_identifier_columns(df)
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:130]: seems like a waste of space, but it's definitely a gotcha reading. changed

File: nwss_wastewater/delphi_nwss/pull.py [hunk 10]
```diff
@@ -34,47 +27,70 @@ def sig_digit_round(value, n_digits):
...
+    specifically, state, and `provider_normalization`, which gives the signal identifier
+    """
+    # a pair of alphanumerics surrounded by _
+    df["state"] = df.key_plot_id.str.extract(r"_(\w\w)_")
+    # anything followed by state ^
+    df["provider"] = df.key_plot_id.str.extract(r"(.*)_[a-z]{2}_")
+    df["signal_name"] = df.provider + "_" + df.normalization
+
+
+def check_endpoints(df):
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:79]: renaming to `check_expected_signals`.

File: nwss_wastewater/delphi_nwss/pull.py [hunk 11]
```diff
@@ -34,47 +27,70 @@ def sig_digit_round(value, n_digits):
...
+def add_identifier_columns(df):
+    """Add identifier columns.
+
+    Add columns to get more detail than key_plot_id gives;
+    specifically, state, and `provider_normalization`, which gives the signal identifier
+    """
+    # a pair of alphanumerics surrounded by _
+    df["state"] = df.key_plot_id.str.extract(r"_(\w\w)_")
+    # anything followed by state ^
+    df["provider"] = df.key_plot_id.str.extract(r"(.*)_[a-z]{2}_")
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:75]: yeah reading it again my text is cryptic. mark as resolved if it's clear enough.

File: nwss_wastewater/delphi_nwss/pull.py [hunk 12]
```diff
@@ -34,47 +27,70 @@ def sig_digit_round(value, n_digits):
...
     df = df.reset_index()
     return df
 
 
-def pull_nwss_data(socrata_token: str):
+def add_identifier_columns(df):
+    """Add identifier columns.
+
+    Add columns to get more detail than key_plot_id gives;
+    specifically, state, and `provider_normalization`, which gives the signal identifier
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/pull.py:70]: got it, leaving this unresolved for feedback

File: nwss_wastewater/delphi_nwss/run.py (hunk 1 above)
Comment (nmdefries) [reply nwss_wastewater/delphi_nwss/run.py:107]: Ah, wastewater treatment plant id seems not worth adding; I don't think it would be particularly useful to users since it's a nonstandard geo and isn't clear how it corresponds to counties, MSAs, etc.

If we were to add wwtp_id, I'd think we'd put it in covidcast, too. We haven't split an indicator across endpoints by geo.

File: nwss_wastewater/delphi_nwss/constants.py [hunk 13]
```diff
@@ -12,18 +12,33 @@
...
+PROVIDER_NORMS = {
+    "provider": ["CDC_VERILY", "CDC_VERILY", "NWSS", "NWSS", "WWS"],
+    "normalization": [
+        "flow-population",
+        "microbial",
+        "flow-population",
+        "microbial",
+        "microbial",
+    ],
 }
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply nwss_wastewater/delphi_nwss/constants.py]: done!

**Comment by dsweber2** [comment]
closing as this is migrating to a new endpoint
